### DISCUSSION TITLE
How to mock time with gocron

### DISCUSSION THREAD

**Discussion by yangzh** [go-co-op/gocron#377] [General]
Hey, maintainer: 

I'm trying to write a scheduler with gocron, and it seems to be serving my needs just fine in production. However, I want to do some testing to ensure my own scheduling logic makes sense and ideally I would like to try mock time: actually the ability to mock time is one of the primary reason I picked gocron in the first place. 

I tried to use `github.com/benbjohnson/clock` to mock time, with [this gist](https://gist.github.com/yangzh/223241281794aa442950fff4013a3a58). Essentially I want to test with the mock time, a scheduler with a periodic job of 2sec, can run (at least) 10 times during a mock period of 30sec.

Ideally I would like the time progression to be invisible by the application code, so that we can play god in the time manipulation. 

However, even the underlying `clock` library seems to do its job, the testing code (the gist) doesn't work as expected. It would be great if you can: 

1.  Take a look at the gist and advise if I missed / overlooked any details from your gocron library;
2. Have a working example of how to use `gocron.TimeWrapper`: currently there is no working example. 

Thanks.
Kevin.

**Comment by JohnRoesler** [comment]
Take a look at https://pkg.go.dev/github.com/go-co-op/gocron#example-Scheduler.CustomTime It was added for this purpose to allow testing with your own logic for time.

**Reply by yangzh** [reply]
Thanks, John: 

I was aware of this, however it's NOT a working example, and I have [the gist](https://gist.github.com/yangzh/223241281794aa442950fff4013a3a58), which does NOT seem to work. 

I need further insight into `gocron` to advise why the gist above didn't work. 
Thanks.

**Reply by JohnRoesler** [reply]
I see your gist now. When I find some time to look into it I will. At first blush it looks reasonable.

**Comment by yangzh** [comment]
I spend some time today digging more, and found a new package for mocking time ("go.chromium.org/luci/common/clock/testclock")

With this new [gist](https://gist.github.com/yangzh/7199b77f5ad64f044deabea87fd00d21), I can have a fake clock that runs faster than wall clock, and it works to some extent. 

See the test result here: 
clock set at 2022-08-23T04:53:07Z
Triggered: 2022-08-23T04:53:07Z (real time=2022-08-23T04:53:07Z)
Triggered: 2022-08-23T04:55:07Z (real time=2022-08-23T04:53:09Z)
Triggered: 2022-08-23T04:57:07Z (real time=2022-08-23T04:53:11Z)
Triggered: 2022-08-23T04:59:07Z (real time=2022-08-23T04:53:13Z)
Triggered: 2022-08-23T05:01:07Z (real time=2022-08-23T04:53:15Z)
Triggered: 2022-08-23T05:03:07Z (real time=2022-08-23T04:53:17Z)
Triggered: 2022-08-23T05:05:07Z (real time=2022-08-23T04:53:19Z)
Triggered: 2022-08-23T05:07:07Z (real time=2022-08-23T04:53:21Z)
Triggered: 2022-08-23T05:09:07Z (real time=2022-08-23T04:53:23Z)
Triggered: 2022-08-23T05:11:08Z (real time=2022-08-23T04:53:25Z)
done in 18.009612458s
PASS

Basically I schedule a job running every 2sec, and inside the box was the mock clock (which runs 60x faster): each trigger reports 2 min inside, the test stops after 10 triggering.

However, ideally I want a slight different idea: I want the scheduler to schedule the events inside the box (following the mock clock). 
If the mock clock runs in a uniform faster fashion (`testclock.FastClock`), the real wall time will be much smaller.
With this, I can test, for example, 10-min scheduling within 10sec. 

Do you think your scheduler package can somehow achieve this? Thanks

**Reply by JohnRoesler** [reply]
I see the root issue of why this isn't working. The custom time will work for doing things like starting a job on a specific date / time and setting the time to that time to make sure it runs.

However, trying accelerate a job scheduled every second doesn't really work because we're using the time package's timer func.

Couple options I see, let me know your thoughts:

1. You could make your job interval configurable and for tests simply change it from 10 min to 1 second for test runs.
2. I implemented essentially that same logic in gocron (locally) by wrapping the call to time.AfterFunc with my own func so that you could override the duration. This would allow you force it to run every second instead of every 10 minutes. In your custom implementation of the AfterFunc you could also change the current time to be 10 minutes in the future. Example below:

```go
package main

import (
	"log"
	"time"

	"github.com/go-co-op/gocron"
)

type myCustomTime struct {
	now *time.Time
}

var _ gocron.TimeWrapper = (*myCustomTime)(nil)

func (m myCustomTime) Now(_ *time.Location) time.Time {
	return *m.now
}

func (m myCustomTime) Sleep(_ time.Duration) {
	time.Sleep(time.Second)
	return
}

func (m myCustomTime) Unix(sec int64, nsec int64) time.Time {
	panic("implement me")
}

func main() {
	now := time.Now()
	mct := myCustomTime{
		now: &now,
	}
	firstTime := true

	s := gocron.NewScheduler(time.UTC)
	s.CustomTime(mct)
	s.CustomTimer(func(d time.Duration, f func()) *time.Timer {
		if firstTime {
			firstTime = false
		} else {
			now = now.Add(time.Minute * 10)
		}
		d = time.Second
		return time.AfterFunc(d, f)
	})

	_, err := s.Every(10).Minutes().Do(func() {
		log.Printf("hello: fake time is %s", now.String())
	})
	if err != nil {
		log.Println(err)
	}

	s.StartAsync()

	time.Sleep(10 * time.Second)
	s.Stop()
}
```

Output:
```
2022/08/23 11:02:12 hello: fake time is 2022-08-23 11:02:12.126785 -0500 CDT m=+0.000312251
2022/08/23 11:02:13 hello: fake time is 2022-08-23 11:12:12.126785 -0500 CDT m=+600.000312251
2022/08/23 11:02:14 hello: fake time is 2022-08-23 11:22:12.126785 -0500 CDT m=+1200.000312251
2022/08/23 11:02:15 hello: fake time is 2022-08-23 11:32:12.126785 -0500 CDT m=+1800.000312251
2022/08/23 11:02:16 hello: fake time is 2022-08-23 11:42:12.126785 -0500 CDT m=+2400.000312251
2022/08/23 11:02:17 hello: fake time is 2022-08-23 11:52:12.126785 -0500 CDT m=+3000.000312251
2022/08/23 11:02:18 hello: fake time is 2022-08-23 12:02:12.126785 -0500 CDT m=+3600.000312251
2022/08/23 11:02:19 hello: fake time is 2022-08-23 12:12:12.126785 -0500 CDT m=+4200.000312251
2022/08/23 11:02:20 hello: fake time is 2022-08-23 12:22:12.126785 -0500 CDT m=+4800.000312251
2022/08/23 11:02:21 hello: fake time is 2022-08-23 12:32:12.126785 -0500 CDT m=+5400.000312251
```

**Reply by yangzh** [reply]
Hey, John:

Thanks for spending time helping out (and I hope the investigation itself was fun). 

Yes, you are right that the current `gocron` package relies on `time` package (instead of abstracting time out) for scheduling, which is the root cause: I don't know the detail, but how much work do you think for an additional layer of abstraction? 

For your proposed options: 
1. this should work, but still not great, as the testing logic was "leaked" into application, and we are NOT playing god. :-)
2. I suspect (without going into their code) your new code shares (at least) some high level idea with the `testclock.FastClock` implementation: I think essentially they speed the flow of time by manually adding more for each elapsed second. However, the difference is that your scheduler can access (and make use) your accelerated time, while (because of hardcoding `time` package), your scheduler won't take the time as reported by `testclock.FastClock`.  Again, it comes to how much work for an abstraction layer of time in your scheduler code. 

Also by the way, I like `testclock.FastClock` better than `testclock.TestClock`, because the former requires much less test/control setup to use, and conceptually simpler. For `testclock.TestClock` and the mock clock from `github.com/benbjohnson/clock`, the test code is still a bit awkward to setup.

I would hope you can think about this proposal and make the extra abstraction layer, which is both nice and elegant. Thanks.

**Reply by JohnRoesler** [reply]
You're welcome.

For now, I can push up what I have for the Timer override. In my code that relies on gocron I don't try and test the bits that actually run the scheduler. Rather I test the function inside the scheduler and call the scheduler from my main.go file and trust that it will run on the interval I set it to. I'm not opposed to adding better ways to do tests that include gocron schedules, but I think there are plenty of ways to work around it such that I'm not motivated to do a lot of work abstracting away the time function. The time.AfterFunc is the last bit that wasn't already abstracted away. Digging into the time.AfterFunc, it looks to be a pretty low level function and I don't see any way to override the time that is used by it.

**Reply by yangzh** [reply]
Thanks.

If you can publish the Timer override (and maybe this particular example of playing with timing), I can probably push this on my end to see if I can get my goal achieved. 

Another thought though: you already provide Sleep override (in your TimeWrapper), is that essentially the same code path as the `AfterFunc` we are discussing here? I want to understand whether `TimeWrapper` is a minimal set of APIs that can achieve maximum customization.

**Reply by JohnRoesler** [reply]
https://pkg.go.dev/time#AfterFunc is the function we're using and not time.Sleep.

**Reply by yangzh** [reply]
then does it make sense to expose `AfterFunc` in `TimeWrapper` instead of `Sleep`?  I know this might be a breaking change, though.

**Reply by JohnRoesler** [reply]
Sleep is called in a few places but not for the regular scheduling of job intervals, hence why it is on the time wrapper interface. 

I added the afterfunc override ability here -> https://github.com/go-co-op/gocron/pull/381/files and the reason it's not on the time interface is that it's not a function on time.Time but rather a separate function and I don't want to force someone who is using the customTime to also have to implement the afterFunc.

**Reply by yangzh** [reply]
Thanks for the work!

I will take a look later today and see how I can achieve my goal on my end. 

Also may I ask when your code can be included in release 1.16.4?  I would assume unless I switch my repo to use the head of `gocron`, a new release of `gocron` package is required, right?

**Reply by JohnRoesler** [reply]
https://github.com/go-co-op/gocron/releases/tag/v1.17.0 there it is

**Reply by yangzh** [reply]
Hey, John: 

I got some time to look into this. Thanks for the code, but it seems I still cannot make the whole thing work.
I think the reason is that even with your `CustomTimer`, it expects a `*time.Timer` (which is really a struct, instead of an interface). 

For other mock clock or `go.chromium.org/luci/common/clock/testclock.FastClock`, typically they will return their own version of `Timer` (again, a concrete struct, instead of an interface), and they are NOT compatible with `*time.Timer`, and there is no easy converters in between. 

To solve this once and for all, we should use an interface for timer and `*time.Timer` is simply an implementation of it: the interface will pretty much have `Reset` and `Stop` API, and that's it. 

Another alternative: similar to your example, is to change the duration on-the-fly. However, this only applies to the `FastClock` case, and even that, `FastClock` doesn't expose the necessary ratio to compute the duration anyway.

It seems tricker than I expected: you really need to abstract out all concrete structs from `time` package.  I will leave that to you, then.

**Comment by wandering-tales** [comment]
> I see the root issue of why this isn't working. The custom time will work for doing things like starting a job on a specific date / time and setting the time to that time to make sure it runs.
> 
> However, trying accelerate a job scheduled every second doesn't really work because we're using the time package's timer func.
> 
> Couple options I see, let me know your thoughts:
> 
>     1. You could make your job interval configurable and for tests simply change it from 10 min to 1 second for test runs.
> 
>     2. I implemented essentially that same logic in gocron (locally) by wrapping the call to time.AfterFunc with my own func so that you could override the duration. This would allow you force it to run every second instead of every 10 minutes. In your custom implementation of the AfterFunc you could also change the current time to be 10 minutes in the future. Example below:
> 
> 
> ```go
> package main
> 
> import (
> 	"log"
> 	"time"
> 
> 	"github.com/go-co-op/gocron"
> )
> 
> type myCustomTime struct {
> 	now *time.Time
> }
> 
> var _ gocron.TimeWrapper = (*myCustomTime)(nil)
> 
> func (m myCustomTime) Now(_ *time.Location) time.Time {
> 	return *m.now
> }
> 
> func (m myCustomTime) Sleep(_ time.Duration) {
> 	time.Sleep(time.Second)
> 	return
> }
> 
> func (m myCustomTime) Unix(sec int64, nsec int64) time.Time {
> 	panic("implement me")
> }
> 
> func main() {
> 	now := time.Now()
> 	mct := myCustomTime{
> 		now: &now,
> 	}
> 	firstTime := true
> 
> 	s := gocron.NewScheduler(time.UTC)
> 	s.CustomTime(mct)
> 	s.CustomTimer(func(d time.Duration, f func()) *time.Timer {
> 		if firstTime {
> 			firstTime = false
> 		} else {
> 			now = now.Add(time.Minute * 10)
> 		}
> 		d = time.Second
> 		return time.AfterFunc(d, f)
> 	})
> 
> 	_, err := s.Every(10).Minutes().Do(func() {
> 		log.Printf("hello: fake time is %s", now.String())
> 	})
> 	if err != nil {
> 		log.Println(err)
> 	}
> 
> 	s.StartAsync()
> 
> 	time.Sleep(10 * time.Second)
> 	s.Stop()
> }
> ```
> 
> Output:
> 
> ```
> 2022/08/23 11:02:12 hello: fake time is 2022-08-23 11:02:12.126785 -0500 CDT m=+0.000312251
> 2022/08/23 11:02:13 hello: fake time is 2022-08-23 11:12:12.126785 -0500 CDT m=+600.000312251
> 2022/08/23 11:02:14 hello: fake time is 2022-08-23 11:22:12.126785 -0500 CDT m=+1200.000312251
> 2022/08/23 11:02:15 hello: fake time is 2022-08-23 11:32:12.126785 -0500 CDT m=+1800.000312251
> 2022/08/23 11:02:16 hello: fake time is 2022-08-23 11:42:12.126785 -0500 CDT m=+2400.000312251
> 2022/08/23 11:02:17 hello: fake time is 2022-08-23 11:52:12.126785 -0500 CDT m=+3000.000312251
> 2022/08/23 11:02:18 hello: fake time is 2022-08-23 12:02:12.126785 -0500 CDT m=+3600.000312251
> 2022/08/23 11:02:19 hello: fake time is 2022-08-23 12:12:12.126785 -0500 CDT m=+4200.000312251
> 2022/08/23 11:02:20 hello: fake time is 2022-08-23 12:22:12.126785 -0500 CDT m=+4800.000312251
> 2022/08/23 11:02:21 hello: fake time is 2022-08-23 12:32:12.126785 -0500 CDT m=+5400.000312251
> ```

This example is not working anymore. Only one line is printed, which means the job is only executed once.
I copied the above example as it is in a new `test.go` in the root of my app folder and run it:

```
/app # go run test.go
2023/10/20 15:22:58 hello: fake time is 2023-10-20 15:22:58.376710844 +0000 UTC m=+0.000135085
/app # ps aux
```

At the time of writing, testing the scheduler for periodical jobs is quite impracticable, as there's no working example anywhere.

**Comment by JohnRoesler** [comment]
Mocking of time isn’t great in v1 of gocron. I’m planning to have a better design in v2 that supports actually mocking and manipulating time. I’m planning to use https://github.com/jonboulle/clockwork, but open to suggestions if there is something better.

**Reply by wandering-tales** [reply]
@JohnRoesler Actually, after some debugging and digging in the `go-cron` source code I have finally managed to make that example work by simply avoiding to treat the first iteration as a special one (the `firstTime` variable):

```golang
package main

import (
	"log"
	"time"

	"github.com/go-co-op/gocron"
)

type myCustomTime struct {
	now *time.Time
}

var _ gocron.TimeWrapper = (*myCustomTime)(nil)

func (m myCustomTime) Now(_ *time.Location) time.Time {
	return *m.now
}

func (m myCustomTime) Sleep(_ time.Duration) {
	time.Sleep(time.Second)
	return
}

func (m myCustomTime) Unix(sec int64, nsec int64) time.Time {
	panic("implement me")
}

func main() {
	now := time.Now()
	mct := myCustomTime{
		now: &now,
	}

	s := gocron.NewScheduler(time.UTC)
	s.CustomTime(mct)
	s.CustomTimer(func(d time.Duration, f func()) *time.Timer {
		now = now.Add(time.Minute * 10)
		d = time.Second
		return time.AfterFunc(d, f)
	})

	_, err := s.Every(10).Minutes().Do(func() {
		log.Printf("hello: fake time is %s", now.String())
	})
	if err != nil {
		log.Println(err)
	}

	s.StartAsync()

	time.Sleep(10 * time.Second)
	s.Stop()
}
```

The reason is in the call to the job [`setTimer`](https://github.com/go-co-op/gocron/blob/cda6e1ca8b6d1487c8ac9b5804fc6868833041a9/scheduler.go#L625C1-L641C1):

```golang
	job.setTimer(s.timer(nr, func() {
		if !next.dateTime.IsZero() {
			for {
				n := s.now().UnixNano() - next.dateTime.UnixNano()
				if n >= 0 {
					break
				}
				select {
				case <-s.executor.ctx.Done():
				case <-time.After(time.Duration(n)):
				}
			}
		}
		s.runContinuous(job)
	}))
}
```

The custom timer function is called for each job execution and the difference between the current time (mocked) and the time of the next scheduled execution should always be a **positive number** in order to exit from loop and perform the re-scheduling.
Avoiding to add the job interval duration to the mocked current time for the first time, always leads to a **negative number** which, in turn causes the `time.After` call to **immediately** exit.

About the last point, it's still not clear to me the need to wait for a "negative time duration", since by design of Golang `time` package, is a no-op.
